tritium-core/src/main/java/com/palantir/tritium/event/Handlers.java

```diff
@@ -0,0 +1,183 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.tritium.event;
+
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.UnsafeArg;
+import com.palantir.tritium.api.event.InstrumentationFilter;
+import java.lang.reflect.Method;
+import java.util.Objects;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class Handlers {
+
+    private static final Logger log = LoggerFactory.getLogger(Handlers.class);
+
+    /**
+     * The caller is expected to check {@link InvocationEventHandler#isEnabled()} prior to calling this method,
+     * allowing argument array allocation to be avoided when the handler is not enabled.
+     */
+    @Nullable
+    public static InvocationContext pre(
+            InvocationEventHandler<?> handler,
+            InstrumentationFilter filter,
+            Object instance,
+            Method method,
+            Object[] args) {
+        try {
+            return filter.shouldInstrument(instance, method, args)
+                    ? handler.preInvocation(instance, method, args)
+                    : DisabledHandlerSentinel.INSTANCE;
+        } catch (Throwable t) {
+            logPreInvocationFailure(handler, instance, method, t);
+            return null;
+        }
+    }
+
+    /**
+     * Identical to {@link #pre(InvocationEventHandler, InstrumentationFilter, Object, Method, Object[])}
+     * except that {@link InvocationEventHandler#isEnabled()} is checked along with
+     * {@link InstrumentationFilter#shouldInstrument(Object, Method, Object[])}. This should be used when
+     * argument array allocation has already occurred and cannot be avoided.
+     */
+    @Nullable
+    public static InvocationContext preWithEnabledCheck(
```

Comment:
Method is very similar to `pre`, they should share a common implementation with additional branches.

I considered always checking `handler.isEnabled()`, but figured since we're doingcodegen, we might as well avoid unnecessary overhead of checking the handler twice.